kasa/smart/modules/firmware.py

```diff
@@ -94,15 +145,60 @@ def update_available(self) -> bool | None:
         """Return True if update is available."""
         if not self._device.is_cloud_connected:
             return None
-        return self.latest_firmware.update_available
+        return self.firmware_update_info.update_available
 
-    async def get_update_state(self):
+    async def get_update_state(self) -> DownloadState:
         """Return update state."""
-        return await self.call("get_fw_download_state")
+        resp = await self.call("get_fw_download_state")
+        state = resp["get_fw_download_state"]
+        return DownloadState(**state)
 
-    async def update(self):
+    async def update(self, progress_cb=None):
         """Update the device firmware."""
-        return await self.call("fw_download")
+        current_fw = self.current_firmware
+        _LOGGER.info(
+            "Going to upgrade from %s to %s",
+            current_fw,
+            self.firmware_update_info.version,
+        )
+        await self.call("fw_download")
+
+        # TODO: read timeout from get_auto_update_info or from get_fw_download_state?
+        async with asyncio_timeout(60 * 5):
+            while True:
+                await asyncio.sleep(0.5)
+                try:
+                    state = await self.get_update_state()
+                except Exception as ex:
+                    _LOGGER.warning(
+                        "Got exception, maybe the device is rebooting? %s", ex
+                    )
+                    continue
+
+                _LOGGER.debug("Update state: %s" % state)
+                if progress_cb is not None:
+                    asyncio.ensure_future(progress_cb(state))
```

Comment:
```suggestion
                    asyncio.create_task(progress_cb(state))
```
This can probably be a `create_task` and maybe we should type hint the callback parameter.

Reply:
Indeed, old habits die hard :-P

Reply:
Done. Added also asserts that check that the callback is really awaited properly.